### PR TITLE
Berkmin counting literals in nogoods responsible for the conflict

### PR DESCRIPTION
**Pull request by rtaupe**
Improvement to BerkMin: Now counting literals in nogoods responsible for the conflict.

### CODE REVIEW

File: src/main/java/at/ac/tuwien/kr/alpha/Util.java [hunk 1]
```diff
@@ -33,4 +58,16 @@
...
 		}
 	}
+
+	/**
+	 * Constructs a {@link Set} containing only the given element.
+	 * 
+	 * @param element
+	 * @return
+	 */
+	public static <E> Set<E> set(E element) {
```
Comment (flowlo) [inline src/main/java/at/ac/tuwien/kr/alpha/Util.java]: There's [`Collections.singleton`](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#singleton-T-) for that.

File: src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java [hunk 2]
```diff
@@ -50,29 +50,33 @@
 public class BerkMin implements BranchingHeuristic {
 	
 	public static final double DEFAULT_ACTIVITY = 0.0;
+	public static final int DEFAULT_SIGN_COUNTER = 0;
```
Comment (flowlo) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java]: The default value for sign counters was used in multiple places. I extracted it as constant to lessen the risk of introducing inconsistencies.

File: src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java [hunk 3]
```diff
@@ -50,29 +50,33 @@
...
+	public static final double DEFAULT_DECAY_FACTOR = 0.25;
+
 	private Assignment assignment;
 	private Map<Integer, Double> activityCounters = new HashMap<>();
 	private Map<Integer, Integer> signCounters = new HashMap<>();
-	private Deque<NoGood> stackOfNoGoods = new ArrayDeque<NoGood>();
-	private int decayAge = 10;
-	private double decayFactor = 0.25;
+	private Deque<NoGood> stackOfNoGoods = new ArrayDeque<>();
+	private int decayAge;
```
Comment (flowlo) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java]: The default value for decay age and decay factor were extracted as they appear to be "magic" constants.

File: src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java [hunk 4]
```diff
@@ -50,29 +50,33 @@
...
-	}
-
-	public BerkMin(Assignment assignment, Predicate<? super Integer> isAtomActiveChoicePoint, int decayAge, double decayFactor) {
-		this(assignment, isAtomActiveChoicePoint);
 		this.decayAge = decayAge;
 		this.decayFactor = decayFactor;
 	}
 
+	public BerkMin(Assignment assignment, Predicate<? super Integer> isAtomActiveChoicePoint) {
+		this(assignment, isAtomActiveChoicePoint, DEFAULT_DECAY_AGE, DEFAULT_DECAY_FACTOR);
```
Comment (flowlo) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java]: The defaulting constructor now does not do anything else except call `this`, which is simpler.

File: src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java [hunk 5]
```diff
@@ -128,22 +132,20 @@ public void newNoGood(NoGood newNoGood) {
 
 	@Override
 	public void newNoGoods(Collection<NoGood> newNoGoods) {
-		for (NoGood noGood : newNoGoods) {
-			newNoGood(noGood);
-		}
+		newNoGoods.forEach(this::newNoGood);
```
Comment (flowlo) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java]: I saw that you are not hostile to using Java 8, so I did that here. Happy to revert this.

File: src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java [hunk 6]
```diff
@@ -128,22 +132,20 @@ public void newNoGood(NoGood newNoGood) {
...
 	}
 
 	@Override
 	public double getActivity(int literal) {
-		Double activity = activityCounters.get(atomOf(literal));
-		return defaultActivityIfNull(activity);
+		return activityCounters.getOrDefault(atomOf(literal), DEFAULT_ACTIVITY);
```
Comment (flowlo) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java]: `Map.getOrDefault` to the rescue!

File: src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java [hunk 7]
```diff
@@ -160,16 +162,17 @@ public int chooseAtom() {
 	
 	@Override
 	public boolean chooseSign(int atom) {
-		if (atom <= 0) {
+		if (!isAtom(atom)) {
```
Comment (flowlo) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java]: We have a utility for this check, that should be used throughout Alpha.

File: src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java [hunk 8]
```diff
@@ -186,34 +189,22 @@ private void pushToStack(NoGood noGood) {
...
 
 	private void incrementActivityCounter(int literal) {
-		int atom = atomOf(literal);
-		Double currentValue = defaultActivityIfNull(activityCounters.get(atom));
-		activityCounters.put(atom, ++currentValue);
-	}
-
-	private double defaultActivityIfNull(Double activity) {
-		return activity == null ? DEFAULT_ACTIVITY : activity;
+		activityCounters.compute(atomOf(literal), (k, v) -> (v == null ? DEFAULT_ACTIVITY : v) + 1);
```
Comment (flowlo) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java]: Again some more Java 8 power.

File: src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java [hunk 9]
```diff
@@ -186,34 +189,22 @@ private void pushToStack(NoGood noGood) {
...
-		signCounters.put(literal, ++currentValue);
+		signCounters.compute(atomOf(literal), (k, v) -> (v == null ? DEFAULT_SIGN_COUNTER : v) + 1);
 	}
 
 	private void decayAllIfTimeHasCome() {
 		stepsSinceLastDecay++;
 		if (stepsSinceLastDecay >= decayAge) {
-			decayAll();
+			// Decay all:
+			activityCounters.replaceAll((k, v) -> v * decayFactor);
```
Comment (flowlo) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java]: As `decayAll` is a expressive one-liner with Java 8 I decided to move it here.

File: src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java [hunk 10]
```diff
@@ -241,24 +232,16 @@ private int getMostActiveChoosableAtom(NoGood noGood) {
...
 	private int getMostActiveChoosableAtom(Stream<Integer> streamOfLiterals) {
-		ActivityComparator activityComparator = new ActivityComparator();
-		Optional<Integer> mostActiveAtom = streamOfLiterals.map(Literals::atomOf).filter(this::isUnassigned).filter(isAtomActiveChoicePoint)
-				.max(activityComparator);
-		return mostActiveAtom.orElse(DEFAULT_CHOICE_ATOM);
+		return streamOfLiterals
+			.map(Literals::atomOf)
+			.filter(this::isUnassigned)
+			.filter(isAtomActiveChoicePoint)
+			.max(Comparator.comparingDouble(this::getActivity))
```
Comment (flowlo) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java]: `Comparator.comparingDouble` comes in very handy and makes `ActivityComparator` redundant.

File: src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java (hunk 5 above)
Comment (rtaupe) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/heuristics/BerkMin.java]: Not hostile at all, just not fluent yet ;-)

**Comment by rtaupe** [comment]
@flowlo Thanks for your improvements.

**Pull request merged**